File: pkg/openwrt/devices.go
package openwrt

import (
	"fmt"
	"strings"

	"github.com/yingxiaomo/homeops/pkg/utils"
	tele "gopkg.in/telebot.v3"
)

func HandleDevices(c tele.Context) error {
	c.Respond(&tele.CallbackResponse{Text: "获取设备列表中..."})

	adg := NewAdGuardClient()
	leases, err := adg.GetDHCPLeases()
	if err == nil && len(leases) > 0 {
		txt := "📱 **当前联网设备 (ADG DHCP)**\n-------------------\n"
		count := 0
		for _, item := range leases {
			if count >= 100 {
				break
			}
			ip, _ := item["ip"].(string)
			if ip == "" {
				ip = "?"
			}
			name, _ := item["hostname"].(string)
			if name == "" {
				name, _ = item["name"].(string)
			}
			if name == "" {
				name = "(未知)"
			}
			mac, _ := item["mac"].(string)
			macStr := ""
			if mac != "" {
				macStr = fmt.Sprintf(" `[%s]`", mac)
			}
			txt += fmt.Sprintf("• %s (%s)%s\n", utils.EscapeMarkdown(name), utils.EscapeMarkdown(ip), macStr)
			count++
		}

		menu := &tele.ReplyMarkup{}
		menu.Inline(
			menu.Row(menu.Data("🔙 返回", "wrt_main")),
		)
		return c.Edit(txt, menu, tele.ModeMarkdown)
	}

	// 2. Fallback to OpenWrt /tmp/dhcp.leases
	res, err := SSHExec("cat /tmp/dhcp.leases")
	if err == nil && strings.TrimSpace(res) != "" {
		txt := "📱 **当前联网设备 (DHCP)**\n-------------------\n"
		lines := strings.Split(res, "\n")
		hasDev := false
		for _, line := range lines {
			parts := strings.Fields(line)
			if len(parts) >= 4 {
				ip := parts[2]
				name := parts[3]
				txt += fmt.Sprintf("• %s (%s)\n", utils.EscapeMarkdown(name), utils.EscapeMarkdown(ip))
				hasDev = true
			}
		}
		if !hasDev {
			txt += "没有活跃设备。"
		}

		menu := &tele.ReplyMarkup{}
		menu.Inline(
			menu.Row(menu.Data("🔙 返回", "wrt_main")),
		)
		return c.Edit(txt, menu, tele.ModeMarkdown)
	}

	// 3. Fallback to ARP
	arp, _ := SSHExec("cat /proc/net/arp")
	if strings.TrimSpace(arp) != "" {
		lines := strings.Split(arp, "\n")
		// Check if there are actual entries (skip header)
		if len(lines) > 1 {
			txt := "📱 **当前邻居列表 (ARP)**\n-------------------\n"
			count := 0
			for i, line := range lines {
				if i == 0 {
					continue
				} // Header
				parts := strings.Fields(line)
				if len(parts) >= 4 {
					ip := parts[0]
					mac := parts[3]
					state := ""
					if len(parts) > 5 {
						state = parts[5]
					}
					txt += fmt.Sprintf("• %s `[%s]` %s\n", ip, mac, state)
					count++
				}
			}
			if count > 0 {
				menu := &tele.ReplyMarkup{}
				menu.Inline(
					menu.Row(menu.Data("🔙 返回", "wrt_main")),
				)
				return c.Edit(txt, menu, tele.ModeMarkdown)
			}
		}
	}

	// 4. Fallback to IP Neigh
	neigh, _ := SSHExec("ip neigh show")
	if strings.TrimSpace(neigh) != "" {
		txt := "📱 **当前邻居列表 (IP Neigh)**\n-------------------\n"
		lines := strings.Split(neigh, "\n")
		count := 0
		for _, line := range lines {
			parts := strings.Fields(line)
			if len(parts) >= 1 {
				ip := parts[0]
				mac := ""
				state := ""
				// format: 192.168.1.135 dev br-lan lladdr 24:18:c6:..:..:.. STALE
				for i, p := range parts {
					if p == "lladdr" && i+1 < len(parts) {
						mac = parts[i+1]
					}
				}
				if len(parts) > 0 {
					state = parts[len(parts)-1]
				}

				macStr := ""
				if mac != "" {
					macStr = fmt.Sprintf(" `[%s]`", mac)
				}
				txt += fmt.Sprintf("• %s%s %s\n", ip, macStr, state)
				count++
			}
		}

		if count > 0 {
			menu := &tele.ReplyMarkup{}
			menu.Inline(
				menu.Row(menu.Data("🔙 返回", "wrt_main")),
			)
			return c.Edit(txt, menu, tele.ModeMarkdown)
		}
	}

	return c.Edit("获取失败或没有活跃设备。", &tele.ReplyMarkup{
		InlineKeyboard: [][]tele.InlineButton{
			{{Text: "🔙 返回", Data: "wrt_main"}},
		},
	})
}

File: plugins/openwrt/devices.py
from telegram import InlineKeyboardButton, InlineKeyboardMarkup
from .connection import ssh_exec
from .helpers import safe_callback_answer
from .adg import get_dhcp_leases

async def wrt_devices(update, context):
    query = update.callback_query
    await safe_callback_answer(query, "获取设备列表中...")
    # Prefer ADGuard Home DHCP leases if available
    adg_leases = await get_dhcp_leases()
    if adg_leases:
        txt = "📱 当前联网设备 (ADG DHCP):\n-------------------\n"
        for it in adg_leases[:100]:
            ip = it.get("ip") or "?"
            name = it.get("name") or "(未知)"
            mac = it.get("mac")
            mac_str = f" [{mac}]" if mac else ""
            txt += f"• {name} ({ip}){mac_str}\n"
        keyboard = [[InlineKeyboardButton("🔙 返回", callback_data="wrt_main")]]
        await query.edit_message_text(txt, reply_markup=InlineKeyboardMarkup(keyboard))
        return
    # Fallback to OpenWrt dhcp leases
    res = ssh_exec("cat /tmp/dhcp.leases")
    if not res or not res.strip():
        arp = ssh_exec("cat /proc/net/arp")
        neigh = ssh_exec("ip neigh show")
        kb = [[InlineKeyboardButton("🔙 返回", callback_data="wrt_main")]]
        if arp and arp.strip():
            txt = "📱 当前邻居列表 (ARP):\n-------------------\n"
            lines = arp.splitlines()[1:]
            count = 0
            for line in lines:
                parts = line.split()
                if len(parts) >= 4:
                    ip = parts[0]
                    mac = parts[3]
                    state = parts[5] if len(parts) > 5 else ""
                    txt += f"• {ip} [{mac}] {state}\n"
                    count += 1
            if count == 0:
                txt += "没有邻居记录。"
            await query.edit_message_text(txt, reply_markup=InlineKeyboardMarkup(kb))
            return
        if neigh and neigh.strip():
            txt = "📱 当前邻居列表 (IP Neigh):\n-------------------\n"
            for line in neigh.splitlines():
                tokens = line.split()
                if not tokens:
                    continue
                ip = tokens[0]
                mac = None
                for i, t in enumerate(tokens):
                    if t == "lladdr" and i + 1 < len(tokens):
                        mac = tokens[i + 1]
                        break
                state = tokens[-1] if tokens else ""
                mac_str = f" [{mac}]" if mac else ""
                txt += f"• {ip}{mac_str} {state}\n"
            await query.edit_message_text(txt, reply_markup=InlineKeyboardMarkup(kb))
            return
        await query.edit_message_text("获取失败。", reply_markup=InlineKeyboardMarkup(kb))
        return
    txt = "📱 当前联网设备 (DHCP):\n-------------------\n"
    lines = res.splitlines()
    for line in lines:
        parts = line.split()
        if len(parts) >= 4:
            ip, name = parts[2], parts[3]
            txt += f"• {name} ({ip})\n"
    if not lines:
        txt += "没有活跃设备。"
    keyboard = [[InlineKeyboardButton("🔙 返回", callback_data="wrt_main")]]
    await query.edit_message_text(txt, reply_markup=InlineKeyboardMarkup(keyboard))
